### PR TITLE
Expose conv2d weight/bias preparation as ops.

### PR DESCRIPTION
**Pull request by LPanosTT**
Issue: https://github.com/tenstorrent/tt-metal/issues/13219

- Created ops that allow on-host conv2d weights and biases to be formatted properly for use with ttnn conv2d.
    - Users can now explicitly format the regular `NHWC` weights into the flattened and padded format required by TTNN

### CODE REVIEW

**Review by sankarmanoj-tt — CHANGES_REQUESTED** [review]
Can we move the new prepare weights/bias functions to another file?

**Review by pavlejosipovic — COMMENTED** [review]
Should we use these metods internally as well?

**Comment by LPanosTT** [comment]
> Should we use these metods internally as well?

Not sure. If It were up to me I'd say it's probably a good idea to phase out the old conv2d and replace its use with this newer, more explicit one. But I'm not on this team so, up to you guys.

**Comment by pavlejosipovic** [comment]
> > Should we use these metods internally as well?
> 
> Not sure. If It were up to me I'd say it's probably a good idea to phase out the old conv2d and replace its use with this newer, more explicit one. But I'm not on this team so, up to you guys.

Maybe I'm not following this well enough, but it seems to me that with your change we get c/p code?

**Comment by LPanosTT** [comment]
> > > Should we use these metods internally as well?
> > 
> > 
> > Not sure. If It were up to me I'd say it's probably a good idea to phase out the old conv2d and replace its use with this newer, more explicit one. But I'm not on this team so, up to you guys.
> 
> Maybe I'm not following this well enough, but it seems to me that with your change we get c/p code?

Yes we do, certainly. I can replace the old weight/bias preparation function with calls to this new one.

**Comment by bbradelTT** [comment]
This change looks a lot bigger than exposing weight/bias preparation as ops.

Why are all the changes and extra code needed?

**Comment by LPanosTT** [comment]
> This change looks a lot bigger than exposing weight/bias preparation as ops.
> 
> Why are all the changes and extra code needed?

The added ops are mostly copied from the old `prepare_conv_weights_biases_and_move_to_device` function. Other than that I'm not sure what you mean by code changes. No logic has been changed. I've pybinded he new op(s) aswell which took a fair number of added lines.

**Comment by bbradelTT** [comment]
> > This change looks a lot bigger than exposing weight/bias preparation as ops.
> > Why are all the changes and extra code needed?
> 
> The added ops are mostly copied from the old `prepare_conv_weights_biases_and_move_to_device` function. Other than that I'm not sure what you mean by code changes. No logic has been changed. I've pybinded he new op(s) aswell which took a fair number of added lines.
> 
> This change is necessary because we want to model the weight preparation in MLIR so that we may push all weights to device before runtime.

conv2d.cpp seems to get 597 new lines. An example of the change is
image

**Comment by LPanosTT** [comment]
@bbradelTT The weight and bias preparation require data held by pre-computed `ParrallelConfig`s, `OptimizedConvBlockConfig`, etc... These were computed in the `conv2d` function and then the specific data was passed to `prepare_conv_weights_biases_and_move_to_device`. These configs are ultimately generated using information about the convolution (dtype, image dimensions, filter dimensions, padding, groups, ...) and the device (grid size, etc). So in order to expose the weight and bias preparation as ops that are user friendly, I'd need to allow them to prepare the weights/bias using the convolution data they already have defined. 

I will do some cleanup to compartmentalize duplicated code that are used by `prepare_conv_bias_for_ttnn` and `prepare_conv_weight_for_ttnn`

**Comment by bbradelTT** [comment]
> @bbradelTT The weight and bias preparation require data held by pre-computed `ParrallelConfig`s, `OptimizedConvBlockConfig`, etc... These were computed in the `conv2d` function and then the specific data was passed to `prepare_conv_weights_biases_and_move_to_device`. These configs are ultimately generated using information about the convolution (dtype, image dimensions, filter dimensions, padding, groups, ...) and the device (grid size, etc). So in order to expose the weight and bias preparation as ops that are user friendly, I'd need to allow them to prepare the weights/bias using the convolution data they already have defined.
> 
> I will do some cleanup to compartmentalize duplicated code that are used by `prepare_conv_bias_for_ttnn` and `prepare_conv_weight_for_ttnn`

Cleanup will be good. Thanks.

**Comment by mywoodstock** [comment]
> > Should we use these metods internally as well?
> 
> Not sure. If It were up to me I'd say it's probably a good idea to phase out the old conv2d and replace its use with this newer, more explicit one. But I'm not on this team so, up to you guys.

Yes we need to -- might be better to do it as a subsequent PR.

**Review by bbradelTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp [hunk 1]
```diff
@@ -650,111 +651,83 @@ void validate_weight_and_bias_tensors(
...
-        }
-    } else {
-        // TODO: fix the need to check this. We should be able to accept any datatype and convert
-        TT_ASSERT(weight_tensor_.get_dtype() == weights_bias_dtype);
-        if (bias_tensor.has_value()) {
-            TT_ASSERT(bias_tensor.value().get_dtype() == weights_bias_dtype);
-        }
-    }
-    weight_tensor_ = ttnn::pad(weight_tensor_, weights_channels_padded_shape.to_array_4D(), tt::tt_metal::Array4D({0, 0, 0, 0}), 0);
+std::tuple<OptimizedConvBlockConfig, TensorMemoryLayout> get_opt_conv_op_block_config_and_shard_layout(
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:654]: is this function still c/p code?

**Review by pavlejosipovic — COMMENTED** [review]
Can you maybe abandon this PR and send new one which doesn't have this clang tidy crap (it's off now)?

**Comment by LPanosTT** [comment]
> Can you maybe abandon this PR and send new one which doesn't have this clang tidy crap (it's off now)?

Sure no problem.

**Pull request closed without merging**